token/core/zkatdlog/crypto/common/schnorr.go

```diff
@@ -10,67 +10,116 @@ import (
 	"github.com/pkg/errors"
 )
 
-// Struct for Schnorr proofs
+// SchnorrProof carries a ZKP for statement (w_1, ..., w_n): Com = \prod_{i=1}^n P_i^w_i
 type SchnorrProof struct {
 	Statement *math.G1
 	Proof     []*math.Zr
 	Challenge *math.Zr
 }
 
-type SchnorrVerifier struct {
-	PedParams []*math.G1
-	Curve     *math.Curve
-}
-
+// SchnorrProver produces a Schnorr proof
 type SchnorrProver struct {
 	*SchnorrVerifier
 	Witness    []*math.Zr
 	Randomness []*math.Zr
 	Challenge  *math.Zr
 }
 
-func (v *SchnorrVerifier) RecomputeCommitment(zkp *SchnorrProof) *math.G1 {
-	com := v.Curve.NewG1()
-	for i, p := range zkp.Proof {
-		com.Add(v.PedParams[i].Mul(p))
-	}
-	com.Sub(zkp.Statement.Mul(zkp.Challenge))
-	return com
-}
-
-func (v *SchnorrVerifier) RecomputeCommitments(zkps []*SchnorrProof, challenge *math.Zr) []*math.G1 {
-	commitments := make([]*math.G1, len(zkps))
-	for i, zkp := range zkps {
-		zkp.Challenge = challenge
-		commitments[i] = v.RecomputeCommitment(zkp)
-	}
-	return commitments
-}
-
-func (v *SchnorrVerifier) ComputeChallenge(pub PublicInput) *math.Zr {
-	raw := pub.Bytes()
-	return v.Curve.HashToZr(raw)
+// SchnorrVerifier verifies a SchnorrProof
+type SchnorrVerifier struct {
+	PedParams []*math.G1
+	Curve     *math.Curve
 }
 
+// Prove produces an array of Zr elements that match the passed
+// challenge, witnesses and randomness.
 func (p *SchnorrProver) Prove() ([]*math.Zr, error) {
+	if p.Curve == nil || p.Curve.GroupOrder == nil {
+		return nil, errors.New("cannot compute Schnorr proof: please initialize curve properly")
+	}
 	if len(p.Witness) != len(p.Randomness) {
-		return nil, errors.Errorf("cannot compute proof")
+		return nil, errors.New("cannot compute Schnorr proof: please initialize witness and randomness correctly")
+	}
+	if p.Challenge == nil {
+		return nil, errors.New("cannot compute Schnorr proof: please initialize challenge")
 	}
 	proof := make([]*math.Zr, len(p.Witness))
+	// p_i = r_i + c*w_i mod q
 	for i := 0; i < len(proof); i++ {
+		if p.Witness[i] == nil || p.Randomness == nil {
+			return nil, errors.New("cannot compute Schnorr proof: please initialize nil elements")
+		}
 		proof[i] = p.Curve.ModMul(p.Challenge, p.Witness[i], p.Curve.GroupOrder)
 		proof[i] = p.Curve.ModAdd(proof[i], p.Randomness[i], p.Curve.GroupOrder)
 	}
 	return proof, nil
 }
 
+// ComputePedersenCommitment returns the commitment to opening relative to base in passed curve.
 func ComputePedersenCommitment(opening []*math.Zr, base []*math.G1, c *math.Curve) (*math.G1, error) {
 	if len(opening) != len(base) {
 		return nil, errors.Errorf("can't compute Pedersen commitment [%d]!=[%d]", len(opening), len(base))
 	}
+	if c == nil {
+		return nil, errors.Errorf("can't compute Pedersen commitment: please initialize curve")
+	}
 	com := c.NewG1()
```

Comment:
does mathlib guarantee that NewG1() is initialized to be the identity element?

Reply:
I believe it is the case for the available implementations.